[Calyx] LICM after Affine ParallelOp unrolling

jiahanxie353 opened the pull request:
This patch does LICM after unrolling `affine.parallel` operations in order to prevent potential memory access contentions.

cgyurgyik commented on lib/Dialect/Calyx/Transforms/AffineParallelUnroll.cpp:
```diff
@@ -33,6 +36,181 @@ using namespace mlir;
+    auto key = std::pair(writeOp.getMemRef(), constIndices);
+    if (constantWriteOpIndices.find(key) != constantWriteOpIndices.end() &&
+        constantWriteOpIndices[key] != parentExecuteRegionOp) {
+      // It cannot occur more than once in different parallel regions, which
```
What is "It"?

cgyurgyik commented on lib/Dialect/Calyx/Transforms/AffineParallelUnroll.cpp:
```diff
@@ -33,6 +36,181 @@ using namespace mlir;
+      continue;
+
+    auto key = std::pair(readOp.getMemRef(), constIndices);
+    if (constantWriteOpIndices.contains(key))
+      continue;
+    constantReadOpCounts[key]++;
```
Would something like this work to avoid double lookup?
```
auto it = constantWriteOpIndices.find(key);
if (it == constantWriteOpIndices.end())
  continue;
++it->second;
```

cgyurgyik commented on test/Dialect/Calyx/affine-parallel-unroll.mlir:
```diff
@@ -180,3 +180,55 @@ module {
+    return
+  }
+}
+
```
I think a couple more simple unit tests would instill some confidence. Like, does this handle cases where a dependent write occurs after in the `affine.for`, e.g.,

```
affine.parallel 0 to 2 {
  %i = affine.for 0 to 8 {
    %1 = store 42, %a[%i];
    %0 = load %a[0];
    ...
```

jiahanxie353 replied:
This example breaks my code, thanks for bringing it up

cgyurgyik replied:
Can we have a test for this?

cgyurgyik replied:
Spoke offline; I'll leave this to your discretion.

cgyurgyik commented on lib/Dialect/Calyx/Transforms/AffineParallelUnroll.cpp:
```diff
@@ -33,6 +36,181 @@ using namespace mlir;
+namespace {
+struct CalyxParLICM {
+  // This pass tries to prevent potential memory banking contention by hoisting
+  // memory accesses after AffineParallelUnroll. It only hoists memory reads
```
You wrote `memory access`, but you mean `memory read` here right?

cgyurgyik reviewed (commented):
A few nits and requests.

jiahanxie353 commented:
I've updated the implementation based on the reviews, thanks @cgyurgyik !
Also, I've renamed `LICM` to `MemoryBankConflictResolver`, which reflects the reality better.

cgyurgyik commented on lib/Dialect/Calyx/Transforms/AffineParallelUnroll.cpp:
```diff
@@ -33,6 +36,224 @@ using namespace mlir;
+    auto key = std::pair(memref, constIndices);
+    // We do not hoist any read as long as it's being written in any parallel
+    // region.
+    auto memrefIsWritten = [&](Value memref) {
+      return llvm::any_of(writtenMemRefs, [memref](const auto &entry) {
+        return entry == memref;
+      });
```
Can this just be `writtenMemRefs.contains(memref)`?
